File: problems/1/twosum_v3.js
/**
 * @param {number[]} array
 * @param {number} target
 * @return {number[]}
 */
var twoSum = function(array, target) {
  const map = new Map();
  const length = array.length;
  for (let i = 0; i < length; i++) {
    const element = target - array[i];
    if (map.has(element)) {
      return [map.get(element), i];
    }
    map.set(array[i], i);
  }
};

console.log(twoSum([2, 7, 11, 15], 26));

File: problems/1/Two_Sum_V3.cs
/// <summary>
/// Complexity Analysis
/// Time Complexity : O(n)
/// Space Complexity: o(n)
/// </summary>

using System.Collections.Generic;
public class Solution {
  public int[] TwoSum (int[] nums, int target) {
    int length = nums.Length;
    Dictionary<int, int> dic = new Dictionary<int, int> ();
    for (int i = 0; i < length; i++) {
      int element = target - nums[i];
      if (dic.ContainsKey (element)) {
        return new int[] { i, dic[element] };
      }
      if (!dic.ContainsKey (nums[i])) {
        dic.Add (nums[i], i);
      }
    }

    return new int[] { };
  }
}
